StreamChatClient_v3/Utils/Timers.swift

```diff
@@ -12,42 +12,58 @@ protocol Timer {
   ///   - timeInterval: The number of seconds after which the timer fires.
   ///   - queue: The queue on which the `onFire` callback is called.
   ///   - onFire: Called when the timer fires.
-  static func schedule(timeInterval: TimeInterval, queue: DispatchQueue, onFire: @escaping () -> Void)
+  /// - Returns: `TimerControl` where you can cancel the timer.
+  @discardableResult
+  static func schedule(timeInterval: TimeInterval, queue: DispatchQueue, onFire: @escaping () -> Void) -> TimerControl
 
   /// Schedules a new repeating timer.
   ///
   /// - Parameters:
   ///   - timeInterval: The number of seconds between timer fires.
   ///   - queue: The queue on which the `onFire` callback is called.
   ///   - onFire: Called when the timer fires.
+  /// - Returns: `RepeatingTimerControl` where you can suspend and resume the timer.
   static func scheduleRepeating(timeInterval: TimeInterval,
                                 queue: DispatchQueue,
-                                onFire: @escaping () -> Void) -> TimerControl
+                                onFire: @escaping () -> Void) -> RepeatingTimerControl
 }
 
 /// Allows resuming and suspending of a timer.
-protocol TimerControl {
+protocol RepeatingTimerControl {
   /// Resumes the timer.
   func resume()
 
   /// Pauses the timer.
   func suspend()
 }
 
+/// Allows cancelling a timer.
+protocol TimerControl {
+  /// Cancels the timer.
+  func cancel()
+}
+
+extension DispatchWorkItem: TimerControl {}
+
 /// Default real-world implementations of timers.
 struct DefaultTimer: Timer {
-  static func schedule(timeInterval: TimeInterval, queue: DispatchQueue, onFire: @escaping () -> Void) {
-    queue.asyncAfter(deadline: .now() + timeInterval, execute: onFire)
+  @discardableResult
+  static func schedule(timeInterval: TimeInterval,
+                       queue: DispatchQueue,
+                       onFire: @escaping () -> Void) -> TimerControl {
+    let worker = DispatchWorkItem(block: onFire)
+    queue.asyncAfter(deadline: .now() + timeInterval, execute: worker)
+    return worker
```

Comment:
Why don't we use `Swift.Timer`?

Reply:
@buh ?

Reply:
>Problem is, Timer requires an active run loop which is not always readily available on background queues. The main thread has an active run loop but this defeats the purpose of having our timer run in the background so its a definite no go.

That's why we use `DispatchSourceTimer` for the repeating timer.

Reply:
for future reference https://medium.com/over-engineering/a-background-repeating-timer-in-swift-412cecfd2ef9